pkg/plugin/connector/cleanup_dispenser.go

```diff
@@ -0,0 +1,82 @@
+// Copyright © 2024 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connector
+
+import (
+	"context"
+
+	"github.com/conduitio/conduit-connector-protocol/pconnector"
+)
+
+// CleanupDispenser dispenses sources and destinations
+// for which a cleanup function will be called after they are torn down.
+type CleanupDispenser struct {
+	Target  Dispenser
+	Cleanup func()
+}
+
+func (c *CleanupDispenser) DispenseSpecifier() (SpecifierPlugin, error) {
+	return c.Target.DispenseSpecifier()
+}
+
+func (c *CleanupDispenser) DispenseSource() (SourcePlugin, error) {
+	plugin, err := c.Target.DispenseSource()
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Maybe I'm paranoid, but if `c.Cleanup` is `nil` this could potentially `panic`. Can we skip wrapping the plugin if that's the case? 🙏 
```suggestion
	if err != nil || c.Cleanup == nil {
		return plugin, err
	}
```

Reply:
Hmm... Maybe not even create a cleanup dispenser if there's no cleanup function?

Reply:
Right, that should be the case, but I was looking at the cleanup dispenser in isolation and saw the potential for a panic when it's not used correctly. It's unlikely to happen, assuming we have the check elsewhere, but I would sleep better if we write code that doesn't make such assumptions 😄

Reply:
It's not quite up to best practices, because it returns an interface instead of a struct, but I'd say it pays out.

Reply:
I'd be up for that, if we rename the method. It's unexpected for a constructor to do this, but not for something like `WithCleanup` or even `OnTeardown`, since the function is called on teardown.

Reply:
Done!